--- exim-4.24/scripts/os-type.no_libc5	2003-11-13 11:09:34.777973822 +0100
+++ exim-4.24/scripts/os-type	2003-11-13 11:09:36.175695694 +0100
@@ -125,16 +125,16 @@
 # easy enough to do a better check, and check the symlink destination or the
 # control file contents and make sure.
 
-Linux)  if [ -L /usr/lib/libc.so ]; then
-            if [ x"$(file /usr/lib/libc.so | grep "libc.so.5")"x != xx ]; then
-                    os=Linux-libc5
-            fi
-        else
-            if grep -q libc.so.5 /usr/lib/libc.so; then
-                    os=Linux-libc5
-            fi
-        fi
-        ;;
+#Linux)  if [ -L /usr/lib/libc.so ]; then
+#            if [ x"$(file /usr/lib/libc.so | grep "libc.so.5")"x != xx ]; then
+#                    os=Linux-libc5
+#            fi
+#        else
+#            if grep -q libc.so.5 /usr/lib/libc.so; then
+#                    os=Linux-libc5
+#            fi
+#        fi
+#        ;;
 
 # In the case of NetBSD we need to distinguish between a.out, ELF
 # and COFF binary formats.  However, a.out and COFF are the same
